Adding interaction callbacks

kdonovan opened the pull request:
Adds callbacks for:
-    onSelect          `// Always called when a file or folder is selected`
-    onSelectFile      `//    Called after onSelect, only on file selection`
-    onSelectFolder    `//    Called after onSelect, only on folder selection`
-    onPreviewOpen     `// File opened`
-    onPreviewClose    `// File closed`
-    onFolderOpen      `// Folder opened`
-    onFolderClose     `// Folder closed`

Allow customizing interactivity of the file browser without requiring overriding the renderers. Specifically for me, this enabled me to trigger loading each subfolder's files dynamically from S3 whenever a folder is opened. I believe the addition of callback hooks will also greatly ease e.g. #35.

marpontes commented:
These events are very useful to the browser when adding behaviour.

There are only a few details:
1. When selecting folders it will return `undefined` if the original key doesn't have a trailing `/` I've figured that out and fixed my source, but it wasn't clear to me as everything was working;
2. Some events like `onCreateFolder` that changes the focus, should trigger these new events. in this example (onSelect, onSelectFolder)

The events matrix should be something like this

| previous events should trigger:        |  onSelect | onSelectFile |  onSelectFolder  |
| ----------------------------------- |:---------:|:------------:|:--------------:|
| onCreateFolder                      |     ✅     |              |        ✅      |
| onCreateFiles                      |     ✅     |       ✅       |              |
| onMoveFolder                      |     ✅     |              |        ✅      |
| onMoveFile                      |     ✅     |         ✅     |              |
| onRenameFolder                      |     ✅     |              |        ✅      |
| onRenameFile                      |     ✅     |         ✅     |              |
| onDeleteFolder                      |     ✅     |              |        ✅      |
| onDeleteFile                      |     ✅     |         ✅     |              |

Thank you for the PR, I am using it into dev already 👍

tabby-or-not commented on src/browser.js:
```diff
@@ -107,6 +107,16 @@ class RawFileBrowser extends React.Component {
 
     icons: {},
+
+    onSelect: (fileOrFolder) => {}, // Always called when a file or folder is selected
```
Can you add prop types for these?

kdonovan replied:
Yep!

tabby-or-not commented on src/browser.js:
```diff
@@ -232,33 +252,41 @@ class RawFileBrowser extends React.Component {
     })
+
+    const selected = this.getFile(key)
+    this.props.onSelect( selected )
```
Should this be called in addition to `onSelectFile` / `onSelectFolder` or as a fallback?

kdonovan replied:
I'd envisioned it as a "this will always fire" default, then also offering more specific "opt in only for files/folders" callbacks (in part because I'm not 100% sure about when the fallback branch would be triggered).

tabby-or-not commented on src/browser.js:
```diff
@@ -232,33 +252,41 @@ class RawFileBrowser extends React.Component {
+    if (this.state.previewFile && this.state.previewFile.key !== file.key) this.closeDetail()
+
+    this.setState({previewFile: file})
+    this.props.onPreviewOpen( file )
```
Consider making this a callback of the setState function?

tabby-or-not commented on src/browser.js:
```diff
@@ -232,33 +252,41 @@ class RawFileBrowser extends React.Component {
-      state.previewFile = null
-      return state
-    })
+    this.props.onPreviewClose( this.state.previewFile )
```
Consider making this a callback of the setState function?

tabby-or-not commented on src/browser.js:
```diff
@@ -268,19 +296,24 @@ class RawFileBrowser extends React.Component {
     this.setState(state => {
       if (folderKey in state.openFolders) {
         delete state.openFolders[folderKey]
+        this.props.onFolderClose( this.getFile(folderKey) )
```
Consider making this a callback of the setState function?

tabby-or-not commented on src/browser.js:
```diff
@@ -268,19 +296,24 @@ class RawFileBrowser extends React.Component {
+        this.props.onFolderClose( this.getFile(folderKey) )
       } else {
         state.openFolders[folderKey] = true
+        this.props.onFolderOpen( this.getFile(folderKey) )
```
Consider making this a callback of the setState function?

tabby-or-not commented on src/browser.js:
```diff
@@ -268,19 +296,24 @@ class RawFileBrowser extends React.Component {
   openFolder = (folderKey) => {
     this.setState(state => {
       state.openFolders[folderKey] = true
+      this.props.onFolderOpen( this.getFile(folderKey) )
```
Consider making this a callback of the setState function?

jarekwg commented:
Thanks!

dguayrobotiq commented:
Any idea when this will be released? I just realized this is not in version `1.4.2`...

tabby-or-not commented:
@dguayrobotiq @kdonovan Released in https://github.com/uptick/react-keyed-file-browser/releases/tag/1.4.3

@kdonovan Do you think you could update the matrix above for the release notes?

kdonovan commented:
@tabby-or-not actually, it looks like that table isn't relevant to what's been merged. 

Unfortunately I haven't used any of the listed events (my use-case was purely read-only display of existing files in S3), so I can't thoroughly review/test all the cases correctly.  It's possible that will change if my client continues down their current route, but otherwise/in the meantime @marpontes table is a great jumping off point for anyone who wants to add a new PR addressing those issues.

dguayrobotiq commented:
Ah thanks guys for the release ! @kdonovan `onSelectFolder` doesn't seem to pass the selected folder as parameter. Value is `undefined`

marpontes commented:
> Ah thanks guys for the release ! @kdonovan `onSelectFolder`, `onFolderOpen` and `onFolderClose` doesn't seem to pass the folder as parameter. Received value is `undefined`.

Hey, @dguayrobotiq I mentioned this issue on my comment. Don't know if that's your case, but if it is, there's a workaround. To make it work you need to make sure that folders have a trailing `/`.

dguayrobotiq commented:
@marpontes ok but how am I suppose to do that? I'm passing the full file path in the `key` attribute like `/my/path/file.jpg`. This library is handling flat path so I'm not sure how I can add a trailing `/`

rejochandran commented:
@dguayrobotiq Guess this is useful only when you do not know the contents of the folder and would like to determine when the folder is opened; only in such a scenario, the fill path will not be mentioned and a trailing `/` is possible. 
However, there can be cases where the folder name has to be known though the file path is pre-determined. this needs to be fixed.
